File: Week_01/java/src/_24_SwapNodesInPairs.java
/**
 * 第一遍：2020/08/14周六 ✅
 * 第二遍：2020/08/14周六
 * 第三遍：2020/05/25周一
 * 第四遍：2020/08/13周四
 * 第五遍：2020/08/14周五
 */
class _24_SwapNodesInPairs {
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode tail = null;
        ListNode ret = head.next;
        while (head != null && head.next != null) {
            ListNode firstNode = head;
            ListNode secondNode = head.next;
            head = secondNode.next;

            if (tail != null) {
                tail.next = secondNode;
            }
            secondNode.next = firstNode;
            firstNode.next = head;
            tail = firstNode;
        }
        return ret;
    }

    class SolutionRecursive {
        public ListNode swapPairs(ListNode head) {
            if (head == null || head.next == null) return head;
            ListNode first = head;
            ListNode second = head.next;
            first.next = swapPairs(second.next);
            second.next = first;
            return second;
        }
    }

    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }
}

File: Week_01/cpp/src/_24_SwapNodesInPairs.cpp
#include <iostream>
using namespace std;
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution
{
public:
    ListNode *swapPairs(ListNode *head)
    {
        ListNode dummy(0, head);
        head = &dummy;
        // 1 -> 2 -> 3 -> 4
        while (head->next && head->next->next)
        {
            auto n = head->next->next;
            head->next->next = n->next;
            n->next = head->next;
            head->next = n;
            head = head->next->next;
        }
        return dummy.next;
    }
};